src/layer/WmtsLayer.ts

```diff
@@ -0,0 +1,104 @@
+import axios, { AxiosRequestConfig } from 'axios';
+
+import { GetMapParams, ApiType } from './const';
+import { AbstractLayer } from './AbstractLayer';
+import { RequestConfiguration } from '../utils/cancelRequests';
+import { ensureTimeout } from '../utils/ensureTimeout';
+import { CACHE_CONFIG_30MIN } from '../utils/cacheHandlers';
+import { getAxiosReqParams } from '../utils/cancelRequests';
+import { bboxToXyz, fetchLayersFromWmtsGetCapabilitiesXml } from './wmts.utils';
+
+interface ConstructorParameters {
+  baseUrl?: string;
+  layerId?: string;
+  title?: string | null;
+  description?: string | null;
+  legendUrl?: string | null;
+  resourceUrl?: string | null;
+}
+
+export class WmtsLayer extends AbstractLayer {
+  protected baseUrl: string;
+  protected layerId: string;
+  protected resourceUrl: string;
+
+  public constructor({
+    baseUrl,
+    layerId,
+    title = null,
+    description = null,
+    legendUrl = null,
+    resourceUrl = null,
+  }: ConstructorParameters) {
+    super({ title, description, legendUrl });
+    this.baseUrl = baseUrl;
+    this.layerId = layerId;
+    this.resourceUrl = resourceUrl;
+  }
+
+  public async updateLayerFromServiceIfNeeded(reqConfig?: RequestConfiguration): Promise<void> {
+    await ensureTimeout(async innerReqConfig => {
+      if (!this.resourceUrl) {
+        const parsedLayers = await fetchLayersFromWmtsGetCapabilitiesXml(this.baseUrl, innerReqConfig);
+        const layer = parsedLayers.find(layerInfo => this.layerId === layerInfo.Name[0]);
+        this.resourceUrl = layer.ResourceUrl;
+      }
+    }, reqConfig);
+  }
+
+  public async getMap(params: GetMapParams, api: ApiType, reqConfig?: RequestConfiguration): Promise<Blob> {
+    return await ensureTimeout(async innerReqConfig => {
+      await this.updateLayerFromServiceIfNeeded(reqConfig);
+      const paramsWithoutEffects = { ...params };
+      delete paramsWithoutEffects.gain;
+      delete paramsWithoutEffects.gamma;
+      delete paramsWithoutEffects.effects;
+      const url = this.getMapUrl(paramsWithoutEffects, api);
+
+      const requestConfig: AxiosRequestConfig = {
+        // 'blob' responseType does not work with Node.js:
+        responseType: typeof window !== 'undefined' && window.Blob ? 'blob' : 'arraybuffer',
+        ...getAxiosReqParams(innerReqConfig, CACHE_CONFIG_30MIN),
+      };
+      const response = await axios.get(url, requestConfig);
+      return response.data;
+    });
+  }
+
+  public getMapUrl(params: GetMapParams, api: ApiType): string {
+    if (api !== ApiType.WMTS) {
+      throw new Error('Only WMTS is supported on this layer');
+    }
+
+    if (!params.bbox && !params.tileCoord) {
+      throw new Error('No bbox or x,y coordinates provided');
+    }
+    if (!this.resourceUrl) {
```

Comment:
If this is not available, we could make the `updateLayerFromServiceIfNeeded` request first, like it's done elsewhere. I'm not a big fan of that because it only leads to bugs imho. Just bringing it up to acknowledge it.

Reply:
This would require us to change `AbstractLayer.getMapUrl` to an async function, which would be a braking change, and not really worth it in my eyes.